[macOS] Darwin build patches

str4d opened the pull request:
Part of #2246.

str4d commented:
This PR will go through several "try-commit" cycles in order for changes here to sequentially "unlock" further issues on the MacOS CI builder, making it progressively "less red" rather than immediately green :smile:

@zkbot try

str4d commented on configure.ac:
```diff
@@ -454,8 +454,14 @@ if test x$use_hardening != xno; then
+  if test x$BUILD_OS = xdarwin; then
+    # Xcode's ld (at least ld64-302.3) doesn't support -z
+    AX_CHECK_LINK_FLAG([[-Wl,-z,relro]], [HARDENED_LDFLAGS="$HARDENED_LDFLAGS -Wl,-z,relro"],[AC_MSG_WARN(Cannot enable RELRO)])
+    AX_CHECK_LINK_FLAG([[-Wl,-z,now]], [HARDENED_LDFLAGS="$HARDENED_LDFLAGS -Wl,-z,now"],[AC_MSG_WARN(Cannot enable BIND_NOW)])
```
Added the security label because these changes need further investigation (perhaps in a separate issue though, if this is merged). It's currently necessary to not require these on Darwin (or clang will not build), but are there other flags we could try instead? Or are these flags maybe not present because they are not necessary? As a datapoint, Bitcoin Core does not require these flags.

str4d commented:
@zkbot try

str4d commented:
@zkbot try

str4d commented:
@zkbot try

str4d commented:
The [next build error](https://ci.z.cash/#/builders/16/builds/95) comes from a warning:
```
In file included from txdb.cpp:6:
In file included from ./txdb.h:9:
In file included from ./coins.h:9:
In file included from ./compressor.h:9:
In file included from ./primitives/transaction.h:20:
In file included from ./zcash/JoinSplit.hpp:8:
./zcash/IncrementalMerkleTree.hpp:132:16: error: instantiation of variable 'libzcash::IncrementalMerkleTree<29, libzcash::SHA256Compress>::emptyroots' required here, but no definition is available [-Werror,-Wundefined-var-template]
        return emptyroots.empty_root(Depth);
               ^
txdb.cpp:76:40: note: in instantiation of member function 'libzcash::IncrementalMerkleTree<29, libzcash::SHA256Compress>::empty_root' requested here
    if (rt == ZCIncrementalMerkleTree::empty_root()) {
                                       ^
./zcash/IncrementalMerkleTree.hpp:140:42: note: forward declaration of template entity is here
    static EmptyMerkleRoots<Depth, Hash> emptyroots;
                                         ^
./zcash/IncrementalMerkleTree.hpp:132:16: note: add an explicit instantiation declaration to suppress this warning if 'libzcash::IncrementalMerkleTree<29, libzcash::SHA256Compress>::emptyroots' is explicitly instantiated in another translation unit
        return emptyroots.empty_root(Depth);
               ^
1 error generated.
```

The corresponding zcash-apple patch [disables the warning](https://github.com/kozyilmaz/zcash-apple/blob/3cfba8192e17ace5372da845fce6fe31fcf0b486/zcash/patches/012-zcash-macos-build-args.patch), which seems to be [a common approach](https://github.com/dealii/dealii/issues/3705). But there may be [a way to refactor this](https://stackoverflow.com/questions/39420569/how-to-fix-a-purported-lack-of-an-explicit-instantiation-declaration-when-comp) to avoid the warning.

str4d commented:
@zkbot try

str4d commented:
@zkbot try

str4d commented:
Test failure above was a transient failure in an unrelated test.

The next bug is in the test suite (and will be fixed by #2806), which means that this PR should now contain fixes for all regular compilation issue on MacOS!

kozyilmaz commented:
This commit https://github.com/str4d/univalue/commit/651aba64cc2a1882bcd56f006bcf2736109fe16c may also be used to fix `Pair`ambiguities instead of `uint64_t` cast

str4d commented:
Yep! I did it in individual commits above though, because I wanted the CI to clearly show the various build errors being fixed. The next step is for reviewers to decide which route they prefer (explicit casting, or implicit conversion inside the UniValue library). Note that the latter would be more preferable if it was upstreamed (so we can keep the library in-sync), which would additionally mean review input from there.

str4d commented:
Rebased on master to fix merge conflict. Still needs review / comment on strategy for addressing `Pair` ambiguities.

braddmiller commented:
@str4d I tried to go through the standard build process on this on MacOS and couldn't get it going. `fetch-params.sh` uses `flock` (line 108) which does not exist in MacOS. 

Running `./zcutil/build.sh` fails with `-bash: ./zcutil/build.sh: No such file or directory`

Are you using a different build process?

kozyilmaz commented:
@braddmiller yes, `flock` does not exists on macOS but there is a separate PR for it: https://github.com/zcash/zcash/pull/2784

Zcash team is in the process of adding macOS support (there are additional PR's already waiting), if you need it now, you can try [this repo](https://github.com/kozyilmaz/zcash-apple) (disclaimer: I maintain that one)

str4d commented:
@braddmiller yes, the regular build process is used by MacOS CI. See https://github.com/zcash/zcash/pull/2674 for the dependencies I found necessary on a Mac Mini.

str4d commented:
Also see #2784 for the flock workaround patch

braddmiller commented:
Running into an error, `consensus/upgrades.cpp:70:1: error: control may reach end of non-void function [-Werror,-Wreturn-type]`

compiler doesn't like the fact that this function may never return.

```
int CurrentEpoch(int nHeight, const Consensus::Params& params) {
     for (auto idxInt = Consensus::MAX_NETWORK_UPGRADES - 1; idxInt >= Consensus::BASE_SPROUT; idxInt--) {
         if (NetworkUpgradeActive(nHeight, params, Consensus::UpgradeIndex(idxInt))) {
             return idxInt;
         }
     }
 }
```

str4d commented:
Aha, that's something new introduced in master (and another reason I want to get this PR merged, so we can start seeing Clang errors in CI). I'll rebase on master and add an end case.

str4d commented:
Just so we have CI proof of the above error:

@zkbot try

str4d commented:
Hmm, the MacOS builder threw this error:

```
ar q libsnark.a libsnark/algebra/curves/alt_bn128/alt_bn128_g1.o libsnark/algebra/curves/alt_bn128/alt_bn128_g2.o libsnark/algebra/curves/alt_bn128/alt_bn128_init.o libsnark/algebra/curves/alt_bn128/alt_bn128_pairing.o libsnark/algebra/curves/alt_bn128/alt_bn128_pp.o libsnark/common/profiling.o libsnark/common/utils.o libsnark/gadgetlib1/constraint_profiling.o
ar: creating archive libsnark.a
if [ -n "" ]; then mkdir -p tmp-ar; cd tmp-ar; for AR_LIB in ; do ar x $AR_LIB; done; ar qc libsnark.a tmp-ar/*; cd ..; rm -r tmp-ar; fi;
ar s libsnark.a
c++ -o libsnark/gtests   libsnark/algebra/curves/tests/test_bilinearity.o libsnark/algebra/curves/tests/test_groups.o libsnark/algebra/fields/tests/test_bigint.o libsnark/algebra/fields/tests/test_fields.o libsnark/gadgetlib1/gadgets/hashes/sha256/tests/test_sha256_gadget.o libsnark/gadgetlib1/gadgets/merkle_tree/tests/test_merkle_tree_gadgets.o libsnark/relations/arithmetic_programs/qap/tests/test_qap.o libsnark/zk_proof_systems/ppzksnark/r1cs_ppzksnark/tests/test_r1cs_ppzksnark.o libsnark/gtests.o libsnark.a -fPIC -DBINARY_OUTPUT -DNO_PT_COMPRESSION=1 -fstack-protector-all -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wno-comment -Wfatal-errors -O2 -march=x86-64 -DMONTGOMERY_OUTPUT -DCURVE_ALT_BN128 -I/Users/zcbbworker/macos-0/macos10-12/build/depends/x86_64-apple-darwin16.7.0/share/../include -Ilibsnark -DNO_PROCPS -static -DSTATIC -L/Users/zcbbworker/macos-0/macos10-12/build/depends/x86_64-apple-darwin16.7.0/share/../lib -Wl,-rpath,/Users/zcbbworker/macos-0/macos10-12/build/depends/x86_64-apple-darwin16.7.0/share/../lib -lgtest -lpthread -lgmpxx -lgmp -lboost_program_options-mt -lsodium
ld: library not found for -lcrt0.o
clang: fatal error: linker command failed with exit code 1 (use -v to see invocation)
make[3]: *** [libsnark/gtests] Error 1
make[2]: *** [snark/libsnark.a] Error 2
```

kozyilmaz commented:
@str4d  "-static" has this affect on macOS, check this patch:
https://github.com/kozyilmaz/zcash-apple/blob/master/zcash/patches/018-zcash-macos-no--static.patch

str4d commented:
Thanks @kozyilmaz!

@zkbot try

str4d commented:
Bug above was fixed, but now Homu is in a state because I tried cancelling the other builders to circle round faster...

@zkbot force

str4d commented:
Okay, next bug was caused by introduction of `z_mergetoaddress`, which has the same `size_t` cast issue as previously seen. Fixed in latest commit.

@zkbot try

str4d commented:
Seems that `TARGET_OS` isn't set in the Makefiles. Trying a different strategy.

@zkbot try

str4d commented:
@zkbot try

str4d commented:
@zkbot try

str4d commented:
Typo.

@zkbot try

kozyilmaz commented:
should I cancel https://github.com/zcash/zcash/pull/2806 after commit https://github.com/zcash/zcash/pull/2820/commits/50e89f3e0ceb05182cb60342ee02cea061d71c97

str4d commented:
@kozyilmaz no, I'd like to merge your PR! But I want to ensure we have a CI trail of all the bugs that are being fixed. I'm continuing this incremental PR through until the entire build process passes (because due to the Make layout and multicore, some of the test code gets compiled before the core code), but once we've seen all the bugs that are in your PR, we can merge yours and I can rebase this on top.

str4d commented:
@zkbot try

str4d commented:
Odd transient failure in one of the RPC tests. Next MacOS failure is another `size_t -> uint64_t` cast.

@zkbot try

str4d commented:
Next failure that appeared was the other one from #2806. I've emulated its fix here so we can see it in CI.

@zkbot try

str4d commented:
Well hey, we're getting into linker errors now! https://ci.z.cash/#/builders/16/builds/183 shows (I've removed some output):

```
/bin/sh ../libtool  --tag=CXX   --mode=link /Users/zcbbworker/macos-0/macos10-12/build/depends/x86_64-apple-darwin16.7.0/share/../native/bin/ccache /Library/Developer/CommandLineTools/usr/bin/clang++ -mmacosx-version-min=10.8 -stdlib=libc++ -std=c++11  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -Werror -fPIE -pipe -O2 -g -fwrapv -fno-strict-aliasing -Qunused-arguments -pthread   -pie -no-undefined  -L/Users/zcbbworker/macos-0/macos10-12/build/depends/x86_64-apple-darwin16.7.0/share/../lib  -Wl,-headerpad_max_install_names -Wl,-dead_strip -o libzcashconsensus.la -rpath /Users/zcbbworker/macos-0/macos10-12/build/depends/x86_64-apple-darwin16.7.0/lib crypto/libzcashconsensus_la-equihash.lo crypto/libzcashconsensus_la-hmac_sha512.lo crypto/libzcashconsensus_la-ripemd160.lo crypto/libzcashconsensus_la-sha1.lo crypto/libzcashconsensus_la-sha256.lo crypto/libzcashconsensus_la-sha512.lo libzcashconsensus_la-hash.lo primitives/libzcashconsensus_la-transaction.lo libzcashconsensus_la-pubkey.lo script/libzcashconsensus_la-zcashconsensus.lo script/libzcashconsensus_la-interpreter.lo script/libzcashconsensus_la-script.lo libzcashconsensus_la-uint256.lo libzcashconsensus_la-utilstrencodings.lo  secp256k1/libsecp256k1.la
libtool: link: /Library/Developer/CommandLineTools/usr/bin/clang++ -mmacosx-version-min=10.8 -stdlib=libc++ -std=c++11 -dynamiclib  -o .libs/libzcashconsensus.0.dylib  crypto/.libs/libzcashconsensus_la-equihash.o crypto/.libs/libzcashconsensus_la-hmac_sha512.o crypto/.libs/libzcashconsensus_la-ripemd160.o crypto/.libs/libzcashconsensus_la-sha1.o crypto/.libs/libzcashconsensus_la-sha256.o crypto/.libs/libzcashconsensus_la-sha512.o .libs/libzcashconsensus_la-hash.o primitives/.libs/libzcashconsensus_la-transaction.o .libs/libzcashconsensus_la-pubkey.o script/.libs/libzcashconsensus_la-zcashconsensus.o script/.libs/libzcashconsensus_la-interpreter.o script/.libs/libzcashconsensus_la-script.o .libs/libzcashconsensus_la-uint256.o .libs/libzcashconsensus_la-utilstrencodings.o   -L/Users/zcbbworker/macos-0/macos10-12/build/depends/x86_64-apple-darwin16.7.0/share/../lib secp256k1/.libs/libsecp256k1.a  -mmacosx-version-min=10.8 -stdlib=libc++ -fstack-protector-all -O2 -g -pthread -Wl,-headerpad_max_install_names -Wl,-dead_strip   -pthread -install_name  /Users/zcbbworker/macos-0/macos10-12/build/depends/x86_64-apple-darwin16.7.0/lib/libzcashconsensus.0.dylib -compatibility_version 1 -current_version 1.0 -Wl,-single_module
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-pthread' [-Wunused-command-line-argument]
Undefined symbols for architecture x86_64:
  "libzcash::JoinSplit<2ul, 2ul>::h_sig(uint256 const&, boost::array<uint256, 2ul> const&, uint256 const&)", referenced from:
      JSDescription::h_sig(libzcash::JoinSplit<2ul, 2ul>&, uint256 const&) const in libzcashconsensus_la-transaction.o
  "boost::system::generic_category()", referenced from:
      __GLOBAL__sub_I_equihash.cpp in libzcashconsensus_la-equihash.o
  "_NetworkUpgradeInfo", referenced from:
      __GLOBAL__sub_I_zcashconsensus.cpp in libzcashconsensus_la-zcashconsensus.o
  "boost::system::system_category()", referenced from:
      __GLOBAL__sub_I_equihash.cpp in libzcashconsensus_la-equihash.o
  "libzcash::Note::Note()", referenced from:
      JSDescription::JSDescription(libzcash::JoinSplit<2ul, 2ul>&, uint256 const&, uint256 const&, boost::array<libzcash::JSInput, 2ul> const&, boost::array<libzcash::JSOutput, 2ul> const&, long long, long long, bool, uint256*) in libzcashconsensus_la-transaction.o
      JSDescription::JSDescription(libzcash::JoinSplit<2ul, 2ul>&, uint256 const&, uint256 const&, boost::array<libzcash::JSInput, 2ul> const&, boost::array<libzcash::JSOutput, 2ul> const&, long long, long long, bool, uint256*) in libzcashconsensus_la-transaction.o
      JSDescription::Randomized(libzcash::JoinSplit<2ul, 2ul>&, uint256 const&, uint256 const&, boost::array<libzcash::JSInput, 2ul>&, boost::array<libzcash::JSOutput, 2ul>&, boost::array<unsigned long, 2ul>&, boost::array<unsigned long, 2ul>&, long long, long long, bool, uint256*, std::__1::function<int (int)>) in libzcashconsensus_la-transaction.o
  "_crypto_generichash_blake2b_final", referenced from:
      GenerateHash(crypto_generichash_blake2b_state const&, unsigned int, unsigned char*, unsigned long) in libzcashconsensus_la-equihash.o
      Equihash<96u, 3u>::IsValidSolution(crypto_generichash_blake2b_state const&, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >) in libzcashconsensus_la-equihash.o
      ...
  "LogAcceptCategory(char const*)", referenced from:
      LogPrint(char const*, char const*) in libzcashconsensus_la-equihash.o
      Equihash<96u, 3u>::IsValidSolution(crypto_generichash_blake2b_state const&, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >) in libzcashconsensus_la-equihash.o
      ...
  "_crypto_generichash_blake2b_update", referenced from:
      GenerateHash(crypto_generichash_blake2b_state const&, unsigned int, unsigned char*, unsigned long) in libzcashconsensus_la-equihash.o
      Equihash<96u, 3u>::IsValidSolution(crypto_generichash_blake2b_state const&, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >) in libzcashconsensus_la-equihash.o
      ...
  "LogPrintStr(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      LogPrint(char const*, char const*) in libzcashconsensus_la-equihash.o
      Equihash<96u, 3u>::IsValidSolution(crypto_generichash_blake2b_state const&, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >) in libzcashconsensus_la-equihash.o
      ...
  "_crypto_generichash_blake2b_init_salt_personal", referenced from:
      Equihash<96u, 3u>::InitialiseState(crypto_generichash_blake2b_state&) in libzcashconsensus_la-equihash.o
      PrecomputedTransactionData::PrecomputedTransactionData(CTransaction const&) in libzcashconsensus_la-interpreter.o
      (anonymous namespace)::GetJoinSplitsHash(CTransaction const&) in libzcashconsensus_la-interpreter.o
      SignatureHash(CScript const&, CTransaction const&, unsigned int, int, long long const&, unsigned int, PrecomputedTransactionData const*) in libzcashconsensus_la-interpreter.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [libzcashconsensus.la] Error 1
```

Looks like some linker flags aren't being set correctly for Darwin...

kozyilmaz commented:
@str4d '--disable-libs' is a workaround for this, but you are right, linker flags should be corrected

str4d commented:
I'm actually on the fence about this now. We *could* link the necessary libraries in, but `libzcashconsensus` has never actually been a supported output (it is in our Gitian build outputs, but we don't bundle it in the Debian packages), and I'm concerned that if we broaden the ABI without care, users might start relying on it (or having issues). I also think it would require more effort to actually be usable beyond v1 transactions, which means post-Overwinter it won't necessarily be useful. I'm wondering whether we should instead disable it by default, but configure our kitchensink builder to build it (so that we still have CI coverage).

str4d commented:
Rebased on master to fix the merge conflict. I also squashed the fixup commits into their corresponding parent commits, given that we have already seen the individual errors be resolved in CI.

mdr0id commented:
Tested ACK:
`./build.sh --disable-rust --disable-libs`

(Note: Used homebrew to pull in missing autotools packages. I can use alternative method to compile these from source, if that process is needed to verify this patch)

braddmiller commented on src/zcash/IncrementalMerkleTree.hpp:
```diff
@@ -21,8 +21,38 @@ class MerklePath {
     inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
-        READWRITE(authentication_path);
-        READWRITE(index);
+        std::vector<std::vector<unsigned char>> pathBytes;
```
Can I get some info on what this function is doing and why it's necessary specifically for the Darwin build?

str4d replied:
The problem is that a `bool` is ambiguous in terms of serialization, at least within the Bitcoin Core serialization framework. On GCC builds, it is serialized as a `uint8_t` (which is very inefficient), but Clang just refuses to compile it.

I have an improved version of this in #3107; I shall rebase this PR to replace the commit implementing this with it.

daira replied:
What's the status of this: can this PR be merged as-is without #3107?

str4d replied:
I pulled in the commit from #3107, so this PR now exactly matches it.

str4d commented:
Rebased the branch to bring in the improved `MerklePath` serialization commit. I also collected the UniValue `uint64_t` commits at the end, so it is easier to replace them if we decide to instead patch UniValue.

str4d commented:
Force-pushed the final three commits to use `static_cast<uint64_t>` instead of `(uint64_t)`.

@zkbot try

str4d commented:
The one failure was transient (ccache tripping up). All other supported builders passed, and the MacOS builder remains at the linker error.

I am inclined to merge this as-is, handling the UniValue casts at their source rather than inside the UniValue code.

braddmiller commented:
I built and ran this on OSX High Sierra earlier this week. Just tried to build it again with --disable-libs flag passed resulting in the following error:

```
clang: error: argument unused during compilation: '-pie' [-Werror,-Wunused-command-line-argument]
make[2]: *** [zcashd] Error 1
make[2]: *** Waiting for unfinished jobs....
mv -f test/.deps/test_bitcoin-accounting_tests.Tpo test/.deps/test_bitcoin-accounting_tests.Po
mv -f wallet/test/.deps/test_test_bitcoin-wallet_tests.Tpo wallet/test/.deps/test_test_bitcoin-wallet_tests.Po
mv -f test/.deps/test_bitcoin-rpc_wallet_tests.Tpo test/.deps/test_bitcoin-rpc_wallet_tests.Po
make[1]: *** [all-recursive] Error 1
make: *** [all-recursive] Error 1
```

str4d commented:
@braddmiller could you please try again with `-j1` and provide a few lines above the error (so we can see what command is triggering it)? I have just run a manual build on our MacOS CI worker with `--disable-libs`, and it appears to build fine (with some warnings). The CI worker is using `clang-900.0.38`.

braddmiller commented:
Command run: `./zcutil/build.sh --disable-rust --disable-libs -j1`

Result:
```
/Library/Developer/CommandLineTools/usr/bin/ranlib libbitcoin_wallet.a
/bin/sh ../libtool  --tag=CXX   --mode=link /Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/share/../native/bin/ccache /Library/Developer/CommandLineTools/usr/bin/clang++ -mmacosx-version-min=10.8 -stdlib=libc++ -std=c++11  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -Werror -fPIE -pipe -O2 -g -fwrapv -fno-strict-aliasing  -pthread   -pie  -L/Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/share/../lib  -Wl,-headerpad_max_install_names -Wl,-dead_strip -o zcashd zcashd-bitcoind.o  libbitcoin_server.a libbitcoin_common.a univalue/libunivalue.la libbitcoin_util.a crypto/libbitcoin_crypto.a libzcash.a snark/libsnark.a ./leveldb/libleveldb.a ./leveldb/libmemenv.a secp256k1/libsecp256k1.la libbitcoin_zmq.a -L/Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib -lzmq -lstdc++ libbitcoin_wallet.a -L/Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/share/../lib -lboost_system -lboost_filesystem -lboost_program_options -lboost_thread -lboost_chrono -ldb_cxx-6.2 -L/Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib -lssl -lcrypto -L/Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib -lcrypto -L/Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib -levent_pthreads -levent -L/Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib -levent crypto/libbitcoin_crypto.a -lgmp -lgmpxx -lboost_system -lcrypto -lsodium
libtool: link: /Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/share/../native/bin/ccache /Library/Developer/CommandLineTools/usr/bin/clang++ -mmacosx-version-min=10.8 -stdlib=libc++ -std=c++11 -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -Werror -fPIE -pipe -O2 -g -fwrapv -fno-strict-aliasing -pthread -pie -Wl,-headerpad_max_install_names -Wl,-dead_strip -o zcashd zcashd-bitcoind.o -Wl,-bind_at_load  -L/Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/share/../lib libbitcoin_server.a libbitcoin_common.a univalue/.libs/libunivalue.a -L/Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.4.0/share/../lib libbitcoin_util.a libzcash.a snark/libsnark.a ./leveldb/libleveldb.a ./leveldb/libmemenv.a secp256k1/.libs/libsecp256k1.a libbitcoin_zmq.a -L/Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib /Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib/libzmq.a -lpthread -lstdc++ libbitcoin_wallet.a -lboost_filesystem -lboost_program_options -lboost_thread -lboost_chrono -ldb_cxx-6.2 -lssl /Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib/libevent_pthreads.a /Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib/libevent.a crypto/libbitcoin_crypto.a /Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib/libgmpxx.a /Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib/libgmp.a -lboost_system -lcrypto /Users/bradleymiller/Development/zcash/depends/x86_64-apple-darwin17.5.0/lib/libsodium.a -pthread
clang: error: argument unused during compilation: '-pie' [-Werror,-Wunused-command-line-argument]
make[2]: *** [zcashd] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all-recursive] Error 1
```

mdr0id commented:
Tested ACK: (see above comment for details related to clang ver)

daira commented on src/wallet/asyncrpcoperation_mergetoaddress.cpp:
```diff
@@ -825,10 +825,10 @@ UniValue AsyncRPCOperation_mergetoaddress::perform_joinsplit(
     UniValue arrOutputMap(UniValue::VARR);
     for (size_t i = 0; i < ZC_NUM_JS_INPUTS; i++) {
-        arrInputMap.push_back(inputMap[i]);
+        arrInputMap.push_back(static_cast<uint64_t>(inputMap[i]));
```
I would mildly prefer the "extend UniValue to support ``size_t``" approach, but I don't object to these casts since we can clean it up later.

str4d commented:
Alright, let's merge this, and address any subsequent MacOS issues in follow-up PRs. Thanks everyone!

@zkbot r+
